File: backend/app/schemas/vision.py
from typing import Annotated
from typing import Any

from pydantic import BaseModel, Field


class BBox(BaseModel):
    x: Annotated[int, Field(ge=0)]
    y: Annotated[int, Field(ge=0)]
    w: Annotated[int, Field(gt=0)]
    h: Annotated[int, Field(gt=0)]
    format: str = "pixel"


class BodyDetection(BaseModel):
    id: int
    class_name: str
    class_id: int
    confidence: Annotated[float, Field(ge=0, le=1)]
    bbox: BBox
    source: str = "mock"
    color_hint: str | None = None
    stable_frames: int = 0
    # Detection ids are frame-local.  A real IFF decision must instead be
    # accumulated against this short-lived, vision-owned body track id.
    track_id: int | None = None
    target_team: str = "unknown"
    range_m: float | None = None
    range_uncertainty_m: float | None = None
    range_calibration_hash: str | None = None
    color_decision: dict[str, Any] | None = None


class BalloonDetection(BaseModel):
    id: int
    confidence: Annotated[float, Field(ge=0, le=1)]
    bbox: BBox
    center_x: int
    center_y: int
    source: str = "mock"


class AimPoint(BaseModel):
    id: int
    x: int
    y: int
    source: str = "mock"


class TrackPlaceholder(BaseModel):
    track_id: int
    detection_id: int
    stable_frames: int


class VisionEvent(BaseModel):
    frame_id: int
    timestamp_ms: int
    source: str
    frame_width: int | None = None
    frame_height: int | None = None
    fps: float
    camera_fps: float | None = None
    detector_fps: float | None = None
    preprocess_ms: float
    inference_ms: float
    postprocess_ms: float
    total_latency_ms: float
    total_ms: float | None = None
    camera_source_kind: str | None = None
    camera_device_path: str | None = None
    frame_origin: str | None = None
    detector_kind: str | None = None
    body_detections: list[BodyDetection]
    balloon_detections: list[BalloonDetection]
    tracks: list[TrackPlaceholder] = Field(default_factory=list)
    aim_points: list[AimPoint] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class BrowserFrameInferenceRequest(BaseModel):
    image_base64: str
    width: Annotated[int, Field(gt=0)] = 1280
    height: Annotated[int, Field(gt=0)] = 720
    device_label: str = "browser_camera"


class CameraStatus(BaseModel):
    camera_mode: str
    source: int | str | None
    connected: bool
    running: bool
    stream_enabled: bool
    width: int
    height: int
    fps: int
    last_error: str | None = None
    selected_device: str | None = None
    selected_backend: str | None = None
    source_mode: str | None = None
    input_format: str | None = None
    resolution: str | None = None
    last_frame_age_ms: int | None = None
    last_capture_error: str | None = None
    is_real_camera_evidence: bool = False
    is_external_usb_camera: bool = False
    is_laptop_camera: bool = False
    hardware_presence_note: str | None = None


class CameraSource(BaseModel):
    id: str
    label: str
    mode: str
    available: bool


class CameraSelectRequest(BaseModel):
    camera_mode: str
    camera_source: int | str | None = None


class VisionConfigUpdate(BaseModel):
    vision_mode: str
    body_model_path: str | None = None
    balloon_model_path: str | None = None
    body_conf_threshold: Annotated[float, Field(ge=0, le=1)] = 0.35
    balloon_conf_threshold: Annotated[float, Field(ge=0, le=1)] = 0.35


class VisionStatus(BaseModel):
    running: bool
    vision_mode: str
    model_loading_required: bool
    body_model_path: str | None
    balloon_model_path: str | None
    body_model_loaded: bool
    balloon_model_loaded: bool
    fps: float
    camera_fps: float | None = None
    detector_fps: float | None = None
    latest_frame_id: int
    latest_latency_ms: float
    latest_total_ms: float | None = None
    camera_source_kind: str | None = None
    frame_origin: str | None = None
    detector_kind: str | None = None
    body_count: int
    balloon_count: int
    warnings: list[str]
    advisory_only: bool = True

File: frontend/src/types/vision.ts
export interface BBox {
  x: number
  y: number
  w: number
  h: number
  format: 'pixel'
}

export interface BodyDetection {
  id: number
  class_name: string
  class_id: number
  confidence: number
  bbox: BBox
  source: string
  color_hint: string | null
  stable_frames: number
  target_team?: string
  range_m?: number | null
  color_decision?: Record<string, unknown> | null
}

export interface BalloonDetection {
  id: number
  confidence: number
  bbox: BBox
  center_x: number
  center_y: number
  source: string
}

export interface AimPoint {
  id: number
  x: number
  y: number
  source: string
}

export interface VisionEvent {
  frame_id: number
  timestamp_ms: number
  source: string
  frame_width?: number | null
  frame_height?: number | null
  fps: number
  camera_fps?: number | null
  detector_fps?: number | null
  preprocess_ms: number
  inference_ms: number
  postprocess_ms: number
  total_latency_ms: number
  total_ms?: number | null
  camera_source_kind?: string | null
  camera_device_path?: string | null
  frame_origin?: string | null
  detector_kind?: string | null
  body_detections: BodyDetection[]
  balloon_detections: BalloonDetection[]
  tracks: Array<{ track_id: number; detection_id: number; stable_frames: number }>
  aim_points: AimPoint[]
  warnings: string[]
}

export interface VisionStatus {
  running: boolean
  vision_mode: string
  model_loading_required: boolean
  body_model_path: string | null
  balloon_model_path: string | null
  body_model_loaded: boolean
  balloon_model_loaded: boolean
  fps: number
  camera_fps?: number | null
  detector_fps?: number | null
  latest_frame_id: number
  latest_latency_ms: number
  latest_total_ms?: number | null
  camera_source_kind?: string | null
  frame_origin?: string | null
  detector_kind?: string | null
  body_count: number
  balloon_count: number
  warnings: string[]
  advisory_only: boolean
}

export interface CameraStatus {
  camera_mode: string
  source: string | number | null
  connected: boolean
  running: boolean
  stream_enabled: boolean
  width: number
  height: number
  fps: number
  last_error: string | null
  selected_device: string | null
  selected_backend: string | null
  source_mode: string | null
  input_format: string | null
  resolution: string | null
  last_frame_age_ms: number | null
  last_capture_error: string | null
  is_real_camera_evidence: boolean
  is_external_usb_camera: boolean
  is_laptop_camera: boolean
  hardware_presence_note: string | null
}

export interface VisionConfig {
  vision_mode: string
  body_model_path: string | null
  balloon_model_path: string | null
  body_conf_threshold: number
  balloon_conf_threshold: number
}

export interface LegacyPerceptionPreset {
  preset_id: string
  source_file: string
  camera_index: number | null
  width: number | null
  height: number | null
  fps: number | null
  color_space: string
  hsv_lower: number[] | number[][] | null
  hsv_upper: number[] | number[][] | null
  blur_kernel: number | number[] | null
  morphology_kernel: number | number[] | null
  min_area: number | null
  max_area: number | null
  circularity_min: number | null
  target_selection_rule: string
  smoothing_enabled: boolean
  kalman_enabled: boolean
  notes: string
  risk_class: string
  advisory_only: boolean
  no_physical_command_generated: boolean
}

export interface LegacyPerceptionPresetList {
  presets: LegacyPerceptionPreset[]
  source_reports: string[]
  forbidden_runtime_tokens_present: boolean
  advisory_only: boolean
  no_physical_command_generated: boolean
}

export interface RealCameraEvidence {
  evidence_id: string
  status: string
  created_at: number
  camera_source: string
  camera_device_path: string | null
  frame_origin: string
  detector: string
  preset_id: string | null
  frame_width: number | null
  frame_height: number | null
  fps_estimate: number | null
  detections_count: number
  target_center_metadata: Record<string, unknown>
  warnings: string[]
  advisory_only: boolean
  no_physical_command_generated: boolean
  physical_command_enabled: boolean
}

export interface RealCameraEvidenceStatus {
  status: string
  camera_source: string
  camera_device_path: string | null
  frame_origin: string
  detector: string
  preset_id: string | null
  frame_width: number | null
  frame_height: number | null
  fps_estimate: number | null
  detections_count: number
  target_center_metadata: Record<string, unknown>
  latest_evidence_id: string | null
  warnings: string[]
  advisory_only: boolean
  no_physical_command_generated: boolean
  physical_command_enabled: boolean
}

export interface CameraHostCommandResult {
  command: string
  status: string
  exit_code: number | null
  output: string
  error: string | null
}

export interface CameraDeviceGroup {
  camera_kind: string
  name: string
  paths: string[]
  preferred_capture_path: string | null
  evidence_status: string
  frame_captured: boolean
  advisory_only: boolean
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}

export interface CameraHostDiagnostic {
  diagnostic_id: string
  created_at: number
  platform: string
  host_camera_devices_detected: boolean
  dev_video_entries: string[]
  camera_groups: CameraDeviceGroup[]
  recommended_usb_device_path: string | null
  selected_camera_device: string | null
  selected_camera_name: string | null
  camera_kind: string
  v4l2_available: boolean
  ffmpeg_available: boolean
  user_in_video_group: boolean
  camera_app_not_seen_note: boolean
  real_camera_capture_attempted: boolean
  real_camera_frame_captured: boolean
  camera_acceptance_status: 'passed' | 'partial' | 'blocked_by_host_os' | 'failed' | string
  blocker_reason: string
  commands: CameraHostCommandResult[]
  suggested_actions: string[]
  advisory_only: boolean
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}

export interface RealCameraAcceptance {
  status: string
  camera_tooling_status: string
  frame_captured: boolean
  device_path: string | null
  width: number | null
  height: number | null
  fps_estimate: number | null
  frame_hash: string | null
  frame_path: string | null
  capture_method: string | null
  selected_camera_device: string | null
  selected_camera_name: string | null
  camera_kind: string
  internal_camera_passed: boolean
  external_usb_camera_passed: boolean
  blocker_reason: string
  camera_host: CameraHostDiagnostic
  latest_evidence: RealCameraEvidence
  advisory_only: boolean
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}

export interface RealCameraSelection {
  selected_camera_device: string
  selected_camera_name: string | null
  camera_kind: string
  advisory_only: boolean
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}
